### DISCUSSION TITLE
Accessing global stiffness and mass matrix for debugging

### DISCUSSION THREAD

**Discussion by adhithya96** [idaholab/moose#27407] [Q&A General] [answered]
Is there a way to access the global jacobian or stiffness matrix and mass matrix in MOOSE ? I am trying to solve eigenvalue problem and I am not getting the results as expected.

**Comment by GiudGiud** [comment, marked as answer]
hello

you can print it with `-ksp_view_pmat` in the petsc options or on the command line
you can also output to a matlab readable format iirc 

Guillaume

**Reply by adhithya96** [reply]
Thanks for the input. I checked petsc manual and found the solution. Adding  ```-mat_view ::ascii_matlab.``` in command line while running the executable works. ```-mat_view :filename.m:ascii_matlab.``` prints the data into a matlab file. 
I was going through the source files of MOOSE to see where I can use std::cout for a few hours

**Reply by adhithya96** [reply]
Is there anyway I can see how the global stiffness matrix  is solved  from the source code file ? Or  are the calls being made to slepc and slepc solves the problem ?

**Reply by GiudGiud** [reply]
Back to your first question, I should know about how to print mass matrix but I dont
@lindsayad if you have time to teach us

You can use the Debug/show_execution_order flag to see a little bit of how the Jacobian is assembled.
Otherwise you need to dig into the code

SLEPc definitely solves eigenvalue problems, with callbacks to moose to compute various quantities like the residual and the Jacobian
Are eigenvalue problems what you are doing?

**Reply by lindsayad** [reply]
In almost all MOOSE simulations only a single matrix is formed for solving the linear problem at each Newton iteration: $J \delta u = -R$ where $J$ is the Jacobian, $\delta u$ is the (candidate) solution update, and $R$ is the residual. We do not separate a mass matrix or stiffness matrix. There are ways you can create those matrices, but before we talk about how to do that, it's probably worthwhile to learn more about what your problem is.

Most people doing eigenvalue calculations in MOOSE do some free power iterations to generate a good initial guess and then do a Newton solve to get the fundamental mode. Is this what you're doing? In such a case, when solving the generalized eigenvalue problem $(Ax) = \lambda (Bx)$, MOOSE provides function evaluations for $Ax$ and $Bx$ and then preconditioning is usually done only with some approximation of $A$ so again only a single matrix data structure is formed

**Reply by adhithya96** [reply]
@lindsayad 
Yes. I am solving eigenvalue problem. I am evaluating the stiffness matrix using ```MatDiffusion``` and  mass matrix using ```CoefReaction```. I have been trying to implement the bloch type periodic boundary condition. Trying to print matrices  to see if I can figure out where I am going wrong.

**Reply by lindsayad** [reply]
periodic boundary conditions introduce constraints that the current eigen classes in MOOSE have problems solving. In order to correctly solve problems with constraints, we have https://github.com/libMesh/libmesh/pull/3821 and https://github.com/idaholab/moose/pull/27276 in progress

**Reply by alimostafavi24** [reply]
> In almost all MOOSE simulations only a single matrix is formed for solving the linear problem at each Newton iteration: J δ u = − R where J is the Jacobian, δ u is the (candidate) solution update, and R is the residual. We do not separate a mass matrix or stiffness matrix. There are ways you can create those matrices, but before we talk about how to do that, it's probably worthwhile to learn more about what your problem is.
> 
> Most people doing eigenvalue calculations in MOOSE do some free power iterations to generate a good initial guess and then do a Newton solve to get the fundamental mode. Is this what you're doing? In such a case, when solving the generalized eigenvalue problem ( A x ) = λ ( B x ) , MOOSE provides function evaluations for A x and B x and then preconditioning is usually done only with some approximation of A so again only a single matrix data structure is formed

Does -mat_view return the sum of the mass and stiffness matrices? What would be the way to obtain them separately?

**Reply by lindsayad** [reply]
It will return the sum. You could explore the [tagging system](https://mooseframework.inl.gov/framework_development/tagging.html) if you want to separate out contributions to the system matrix
